subprojects/kotlin-dsl-tooling-builders/src/crossVersionTest/groovy/org/gradle/kotlin/dsl/tooling/builders/r54/KotlinBuildScriptModelCrossVersionSpec.groovy

```diff
@@ -0,0 +1,574 @@
+
+        given:
+        withDefaultSettings()
+
+        def scriptPluginDependency =
+            withFile("script-plugin-dependency.jar")
+
+        def scriptPlugin = withFile("plugin.gradle.kts", """
+            buildscript {
+                dependencies { classpath(files("${scriptPluginDependency.name}")) }
+            }
+
+            // Shouldn't be evaluated
+            throw IllegalStateException()
+        """)
+
+        when:
+        def model = kotlinBuildScriptModelFor(projectDir, scriptPlugin)
+
+        then:
+        assertThat(
+            "Script body shouldn't be evaluated",
+            model.exceptions,
+            equalTo([]))
+
+        and:
+        def scriptPluginClassPath = canonicalClasspathOf(model)
+        assertThat(
+            scriptPluginClassPath.collect { it.name } as List<String>,
+            hasItem(scriptPluginDependency.name))
+
+        assertContainsGradleKotlinDslJars(scriptPluginClassPath)
+    }
+
+    def "sourcePath includes Gradle sources"() {
+
+        expect:
+        assertSourcePathIncludesGradleSourcesGiven("", "")
+    }
+
+    def "sourcePath includes kotlin-stdlib sources resolved against project"() {
+
+        expect:
+        assertSourcePathIncludesKotlinStdlibSourcesGiven(
+            "",
+            "buildscript { $repositoriesBlock }"
+        )
+    }
+
+    def "sourcePath includes kotlin-stdlib sources resolved against project hierarchy"() {
+
+        expect:
+        assertSourcePathIncludesKotlinStdlibSourcesGiven(
+            "buildscript { $repositoriesBlock }",
+            ""
+        )
+    }
+
+    def "sourcePath includes buildscript classpath sources resolved against project"() {
+
+        expect:
+        assertSourcePathIncludesKotlinPluginSourcesGiven(
+            "",
+            """
+                buildscript {
+                    dependencies { classpath(embeddedKotlin("gradle-plugin")) }
+                    $repositoriesBlock
+                }
+            """
+        )
+    }
+
+    def "sourcePath includes buildscript classpath sources resolved against project hierarchy"() {
+
+        expect:
+        assertSourcePathIncludesKotlinPluginSourcesGiven(
+            """
+                buildscript {
+                    dependencies { classpath(embeddedKotlin("gradle-plugin")) }
+                    $repositoriesBlock
+                }
+            """,
+            ""
+        )
+    }
+
+    def "sourcePath includes plugins classpath sources resolved against project"() {
+
+        expect:
+        assertSourcePathIncludesKotlinPluginSourcesGiven(
+            "",
+            """ plugins { kotlin("jvm") version "$targetKotlinVersion" } """
+        )
+    }
+
+    @LeaksFileHandles("Kotlin compiler daemon on buildSrc jar")
+    def "sourcePath includes buildSrc source roots"() {
+
+        given:
+        withKotlinBuildSrc()
+        withSettings("""include(":sub")""")
```

Comment:
:x: Better to augment the default settings like we do in other places, `withDefaultSettings().appendText(...)`